### ISSUE TITLE
[Bug] Exception: Cannot find valid image after 1000! Please check your image path and pipeline

### ISSUE THREAD

**Issue by ryanalexmartin** [open-mmlab/mmdetection#9741]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmdetection/issues) and [Discussions](https://github.com/open-mmlab/mmdetection/discussions) but cannot get the expected help.
- [X] I have read the [FAQ documentation](https://mmdetection.readthedocs.io/en/latest/faq.html) but cannot get the expected help.
- [X] The bug has not been fixed in the [latest version (master)](https://github.com/open-mmlab/mmdetection) or [latest version (3.x)](https://github.com/open-mmlab/mmdetection/tree/dev-3.x).

### Task

I have modified the scripts/configs, or I'm working on my own tasks/models/datasets.

### Branch

master branch https://github.com/open-mmlab/mmdetection

### Environment

sys.platform: linux
Python: 3.9.14 (main, Nov 22 2022, 07:09:51) [GCC 9.4.0]
CUDA available: True
numpy_random_seed: 2147483648
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA RTX A4000
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 11.3, V11.3.109
GCC: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
PyTorch: 1.10.2
PyTorch compiling details: PyTorch built with:
  - GCC 9.4
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_60,code=compute_60;-gencode;arch=compute_70,code=compute_70;-gencode;arch=compute_80,code=compute_80
  - CuDNN 8.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/usr/bin/c++, CXX_FLAGS= -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.2, USE_CUDA=1, USE_CUDNN=1, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

TorchVision: 0.11.0a0+05eae32
OpenCV: 4.6.0
MMEngine: 0.5.0
MMDetection: 3.0.0rc5+

### Reproduces the problem - code sample

```# dataset settings
dataset_type = 'CocoDataset'
data_root = '/nazar-mmdetection/mmdetection/bigdata/' # I've tried both SMB and local storage

file_client_args = dict(backend='disk')

train_pipeline = [
    dict(type='LoadImageFromFile', file_client_args=file_client_args),
    dict(type='LoadAnnotations', with_bbox=True, with_mask=True, poly2mask=False),
    dict(type='Resize', scale=(1333, 800), keep_ratio=True),
    dict(type='RandomFlip', prob=0.5),
    dict(type='PackDetInputs')
]
test_pipeline = [
    dict(type='LoadImageFromFile', file_client_args=file_client_args),
    dict(type='Resize', scale=(1333, 800), keep_ratio=True),
    # If you don't have a gt annotation, delete the pipeline
    dict(type='LoadAnnotations', with_bbox=True, with_mask=True, poly2mask=False),
    dict(
        type='PackDetInputs',
        meta_keys=('img_id', 'img_path', 'ori_shape', 'img_shape',
                   'scale_factor'))
]
train_dataloader = dict(
    batch_size=2,
    num_workers=32,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=False),
    batch_sampler=dict(type='AspectRatioBatchSampler'),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file='annotations/instances_Train.json',
        data_prefix=dict(img='images/'),
        filter_cfg=dict(filter_empty_gt=False, min_size=32),
        pipeline=train_pipeline))
val_dataloader = dict(
    batch_size=1,
    num_workers=2,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file='annotations/instances_Validation.json',
        data_prefix=dict(img='images/'),
        test_mode=True,
        pipeline=test_pipeline))
test_dataloader = val_dataloader

val_evaluator = dict(
    type='CocoMetric',
    ann_file=data_root + 'annotations/instances_Validation.json',
    metric=['bbox', 'segm'],
    format_only=False)
test_evaluator = val_evaluator

```

### Reproduces the problem - command or script

Both the `tools/train.py` and `tools/analysis_tools/browse_dataset.py` give this error.

### Reproduces the problem - error message

```Original Traceback (most recent call last):
  File "/python/lib/python3.9/site-packages/torch/utils/data/_utils/worker.py", line 287, in _worker_loop
    data = fetcher.fetch(index)
  File "/python/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 49, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/python/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 49, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/nazar-mmdetection/mmengine/mmengine/dataset/base_dataset.py", line 416, in __getitem__
    raise Exception(f'Cannot find valid image after {self.max_refetch}! '
Exception: Cannot find valid image after 1000! Please check your image path and pipeline
```

### Additional information

Because my classes weren't specified, for some reason it wasn't able to find any of the images.

```
train_dataloader = dict(
    batch_size=2,
    num_workers=2,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=False),
    batch_sampler=dict(type='AspectRatioBatchSampler'),
    dataset=dict(
        type=dataset_type,

        metainfo=dict(classes=classes), # I HAD TO ADD THIS LINE

        data_root=data_root,
        ann_file='annotations/instances_Train.json',
        data_prefix=dict(img='images'),
        filter_cfg=dict(filter_empty_gt=False, min_size=32),
        pipeline=train_pipeline))
```
        
Adding the above line, _for some reason_, was enough to let my dataloader find the correct images.  I think this is an example of failing silently.  I'm not sure why a missing class definition would make it impossible for the dataloader to find the images.

Maybe we should add a check to the dataloader to make sure that the classes are defined?

**Comment by lxiaoxiaoxing** [comment]
It seems that your custom image size is too small

**Comment by convergence-lab** [comment]
I met the same problem. In my case, I comment out some data check code from my custom dataset class.
My custom dataset is made by COCOAnnotator. it is not pass the test.

```
    def parse_data_info(self, raw_data_info: dict) -> Union[dict, List[dict]]:
        """Parse raw annotation to target format.

        Args:
            raw_data_info (dict): Raw data information load from ``ann_file``

        Returns:
            Union[dict, List[dict]]: Parsed annotation.
        """
        img_info = raw_data_info["raw_img_info"]
        ann_info = raw_data_info["raw_ann_info"]

        data_info = {}

        # TODO: need to change data_prefix['img'] to data_prefix['img_path']
        img_path = osp.join(self.data_prefix["img"], img_info["file_name"])
        if self.data_prefix.get("seg", None):
            seg_map_path = osp.join(
                self.data_prefix["seg"],
                img_info["file_name"].rsplit(".", 1)[0] + self.seg_map_suffix,
            )
        else:
            seg_map_path = None
        data_info["img_path"] = img_path
        data_info["img_id"] = img_info["img_id"]
        data_info["seg_map_path"] = seg_map_path
        data_info["height"] = img_info["height"]
        data_info["width"] = img_info["width"]

        if self.return_classes:
            data_info["text"] = self.metainfo["classes"]
            data_info["custom_entities"] = True

        instances = []
        for i, ann in enumerate(ann_info):
            instance = {}

            if ann.get("ignore", False):
                continue
            x1, y1, w, h = ann["bbox"]
            # inter_w = max(0, min(x1 + w, img_info["width"]) - max(x1, 0))
            # inter_h = max(0, min(y1 + h, img_info["height"]) - max(y1, 0))
            # if inter_w * inter_h == 0:
            # continue
            # if ann["area"] <= 0 or w < 1 or h < 1:
            # continue
            # if ann["category_id"] not in self.cat_ids:
            # continue
            bbox = [x1, y1, x1 + w, y1 + h]

            if ann.get("iscrowd", False):
                instance["ignore_flag"] = 1
            else
                instance["ignore_flag"] = 0
            instance["bbox"] = bbox
            instance["bbox_label"] = self.cat2label[ann["category_id"]]

            if ann.get("segmentation", None):
                instance["mask"] = ann["segmentation"]

            instances.append(instance)
        data_info["instances"] = instances

        return data_info
```

**Comment by AnasTrace** [comment]
> I met the same problem. In my case, I comment out some data check code from my custom dataset class. My custom dataset is made by COCOAnnotator. it is not pass the test.
> 
> ```
>     def parse_data_info(self, raw_data_info: dict) -> Union[dict, List[dict]]:
>         """Parse raw annotation to target format.
> 
>         Args:
>             raw_data_info (dict): Raw data information load from ``ann_file``
> 
>         Returns:
>             Union[dict, List[dict]]: Parsed annotation.
>         """
>         img_info = raw_data_info["raw_img_info"]
>         ann_info = raw_data_info["raw_ann_info"]
> 
>         data_info = {}
> 
>         # TODO: need to change data_prefix['img'] to data_prefix['img_path']
>         img_path = osp.join(self.data_prefix["img"], img_info["file_name"])
>         if self.data_prefix.get("seg", None):
>             seg_map_path = osp.join(
>                 self.data_prefix["seg"],
>                 img_info["file_name"].rsplit(".", 1)[0] + self.seg_map_suffix,
>             )
>         else:
>             seg_map_path = None
>         data_info["img_path"] = img_path
>         data_info["img_id"] = img_info["img_id"]
>         data_info["seg_map_path"] = seg_map_path
>         data_info["height"] = img_info["height"]
>         data_info["width"] = img_info["width"]
> 
>         if self.return_classes:
>             data_info["text"] = self.metainfo["classes"]
>             data_info["custom_entities"] = True
> 
>         instances = []
>         for i, ann in enumerate(ann_info):
>             instance = {}
> 
>             if ann.get("ignore", False):
>                 continue
>             x1, y1, w, h = ann["bbox"]
>             # inter_w = max(0, min(x1 + w, img_info["width"]) - max(x1, 0))
>             # inter_h = max(0, min(y1 + h, img_info["height"]) - max(y1, 0))
>             # if inter_w * inter_h == 0:
>             # continue
>             # if ann["area"] <= 0 or w < 1 or h < 1:
>             # continue
>             # if ann["category_id"] not in self.cat_ids:
>             # continue
>             bbox = [x1, y1, x1 + w, y1 + h]
> 
>             if ann.get("iscrowd", False):
>                 instance["ignore_flag"] = 1
>             else
>                 instance["ignore_flag"] = 0
>             instance["bbox"] = bbox
>             instance["bbox_label"] = self.cat2label[ann["category_id"]]
> 
>             if ann.get("segmentation", None):
>                 instance["mask"] = ann["segmentation"]
> 
>             instances.append(instance)
>         data_info["instances"] = instances
> 
>         return data_info
> ```

It works for me to

**Comment by ljh-7** [comment]
> It seems that the image is not find. You might need to check the path of your data.

so what should I do to solve this problem? It seems that my custom image size is too small?

**Comment by ljh-7** [comment]
> ### Prerequisite
> * [x]  I have searched [Issues](https://github.com/open-mmlab/mmdetection/issues) and [Discussions](https://github.com/open-mmlab/mmdetection/discussions) but cannot get the expected help.
> * [x]  I have read the [FAQ documentation](https://mmdetection.readthedocs.io/en/latest/faq.html) but cannot get the expected help.
> * [x]  The bug has not been fixed in the [latest version (master)](https://github.com/open-mmlab/mmdetection) or [latest version (3.x)](https://github.com/open-mmlab/mmdetection/tree/dev-3.x).
> 
> ### Task
> I have modified the scripts/configs, or I'm working on my own tasks/models/datasets.
> 
> ### Branch
> master branch https://github.com/open-mmlab/mmdetection
> 
> ### Environment
> sys.platform: linux Python: 3.9.14 (main, Nov 22 2022, 07:09:51) [GCC 9.4.0] CUDA available: True numpy_random_seed: 2147483648 GPU 0: NVIDIA GeForce RTX 3090 GPU 1: NVIDIA RTX A4000 CUDA_HOME: /usr/local/cuda NVCC: Cuda compilation tools, release 11.3, V11.3.109 GCC: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0 PyTorch: 1.10.2 PyTorch compiling details: PyTorch built with:
> 
> * GCC 9.4
> * C++ Version: 201402
> * Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
> * OpenMP 201511 (a.k.a. OpenMP 4.5)
> * LAPACK is enabled (usually provided by MKL)
> * NNPACK is enabled
> * CPU capability usage: AVX2
> * CUDA Runtime 11.3
> * NVCC architecture flags: -gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_60,code=compute_60;-gencode;arch=compute_70,code=compute_70;-gencode;arch=compute_80,code=compute_80
> * CuDNN 8.2
> * Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/usr/bin/c++, CXX_FLAGS= -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.2, USE_CUDA=1, USE_CUDNN=1, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON,
> 
> TorchVision: 0.11.0a0+05eae32 OpenCV: 4.6.0 MMEngine: 0.5.0 MMDetection: 3.0.0rc5+
> 
> ### Reproduces the problem - code sample
> ```
> dataset_type = 'CocoDataset'
> data_root = '/nazar-mmdetection/mmdetection/bigdata/' # I've tried both SMB and local storage
> 
> file_client_args = dict(backend='disk')
> 
> train_pipeline = [
>     dict(type='LoadImageFromFile', file_client_args=file_client_args),
>     dict(type='LoadAnnotations', with_bbox=True, with_mask=True, poly2mask=False),
>     dict(type='Resize', scale=(1333, 800), keep_ratio=True),
>     dict(type='RandomFlip', prob=0.5),
>     dict(type='PackDetInputs')
> ]
> test_pipeline = [
>     dict(type='LoadImageFromFile', file_client_args=file_client_args),
>     dict(type='Resize', scale=(1333, 800), keep_ratio=True),
>     # If you don't have a gt annotation, delete the pipeline
>     dict(type='LoadAnnotations', with_bbox=True, with_mask=True, poly2mask=False),
>     dict(
>         type='PackDetInputs',
>         meta_keys=('img_id', 'img_path', 'ori_shape', 'img_shape',
>                    'scale_factor'))
> ]
> train_dataloader = dict(
>     batch_size=2,
>     num_workers=32,
>     persistent_workers=True,
>     sampler=dict(type='DefaultSampler', shuffle=False),
>     batch_sampler=dict(type='AspectRatioBatchSampler'),
>     dataset=dict(
>         type=dataset_type,
>         data_root=data_root,
>         ann_file='annotations/instances_Train.json',
>         data_prefix=dict(img='images/'),
>         filter_cfg=dict(filter_empty_gt=False, min_size=32),
>         pipeline=train_pipeline))
> val_dataloader = dict(
>     batch_size=1,
>     num_workers=2,
>     persistent_workers=True,
>     drop_last=False,
>     sampler=dict(type='DefaultSampler', shuffle=False),
>     dataset=dict(
>         type=dataset_type,
>         data_root=data_root,
>         ann_file='annotations/instances_Validation.json',
>         data_prefix=dict(img='images/'),
>         test_mode=True,
>         pipeline=test_pipeline))
> test_dataloader = val_dataloader
> 
> val_evaluator = dict(
>     type='CocoMetric',
>     ann_file=data_root + 'annotations/instances_Validation.json',
>     metric=['bbox', 'segm'],
>     format_only=False)
> test_evaluator = val_evaluator
> ```
> 
> ### Reproduces the problem - command or script
> Both the `tools/train.py` and `tools/analysis_tools/browse_dataset.py` give this error.
> 
> ### Reproduces the problem - error message
> ```
>   File "/python/lib/python3.9/site-packages/torch/utils/data/_utils/worker.py", line 287, in _worker_loop
>     data = fetcher.fetch(index)
>   File "/python/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 49, in fetch
>     data = [self.dataset[idx] for idx in possibly_batched_index]
>   File "/python/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 49, in <listcomp>
>     data = [self.dataset[idx] for idx in possibly_batched_index]
>   File "/nazar-mmdetection/mmengine/mmengine/dataset/base_dataset.py", line 416, in __getitem__
>     raise Exception(f'Cannot find valid image after {self.max_refetch}! '
> Exception: Cannot find valid image after 1000! Please check your image path and pipeline
> ```
> 
> ### Additional information
> Because my classes weren't specified, for some reason it wasn't able to find any of the images.
> 
> ```
> train_dataloader = dict(
>     batch_size=2,
>     num_workers=2,
>     persistent_workers=True,
>     sampler=dict(type='DefaultSampler', shuffle=False),
>     batch_sampler=dict(type='AspectRatioBatchSampler'),
>     dataset=dict(
>         type=dataset_type,
> 
>         metainfo=dict(classes=classes), # I HAD TO ADD THIS LINE
> 
>         data_root=data_root,
>         ann_file='annotations/instances_Train.json',
>         data_prefix=dict(img='images'),
>         filter_cfg=dict(filter_empty_gt=False, min_size=32),
>         pipeline=train_pipeline))
> ```
> 
> Adding the above line, _for some reason_, was enough to let my dataloader find the correct images. I think this is an example of failing silently. I'm not sure why a missing class definition would make it impossible for the dataloader to find the images.
> 
> Maybe we should add a check to the dataloader to make sure that the classes are defined?

do you solve it?

**Comment by Trotts** [comment]
Hitting the same issue, is there a fix?
